src/utils/crypto_utils/src/crypto_lib.c

```diff
@@ -375,3 +378,782 @@ CryptoKeyHandle GetRootKeyForKeyID(const char* kid)
+    {
+        goto done;
+    }
+
+    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
+
+    if (ctx == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    const EVP_CIPHER* cipherType = CryptoUtils_GetCipherType(alg);
+
+    if (cipherType == UNSUPPORTED_DECRYPTION_ALG)
+    {
+        result.ExtendedResultCode = ADUC_ERC_CRYPTO_ALG_NOTSUPPORTED;
+        goto done;
+    }
+
+    const size_t blockSize = EVP_CIPHER_block_size(cipherType);
+
+    if (blockSize != plainTextBlockSize)
+    {
+        result.ExtendedResultCode = ADUC_ERC_INCORRECT_BLOCK_SIZE;
+        goto done;
+    }
+
+    const CONSTBUFFER* keyData = CONSTBUFFER_GetContent(eKey->keyData);
+    if (EVP_EncryptInit_ex(ctx, cipherType, NULL, keyData->buffer, iv) != 1)
+    {
+        result.ExtendedResultCode = ADUC_ERC_ENCRYPTION_INIT_FAILED;
+        goto done;
+    }
+
+    if (EVP_CIPHER_CTX_set_padding(ctx, 0) != 1)
+    {
+        goto done;
+    }
+
+    size_t outSize = 0;
+
+    encryptedBuffer = (unsigned char*)malloc(blockSize);
+
+    if (encryptedBuffer == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    int encryptedLength = 0;
+    if (EVP_EncryptUpdate(ctx, encryptedBuffer, &encryptedLength, block, plainTextBlockSize) != 1)
+    {
+        result.ExtendedResultCode = ADUC_ERC_ENCRYPTION_UPDATE_FAILED;
+        goto done;
+    }
+
+    outSize += encryptedLength;
+
+    if (outSize > plainTextBlockSize)
+    {
+        result.ExtendedResultCode = ADUC_ERC_INCORRECT_BLOCK_SIZE;
+        goto done;
+    }
+
+    int finalEncryptedLength = 0;
+
+    if (EVP_EncryptFinal_ex(ctx, encryptedBuffer + encryptedLength, &finalEncryptedLength) != 1)
+    {
+        result.ExtendedResultCode = ADUC_ERC_ENCRYPTION_FINAL_FAILED;
+        goto done;
+    }
+
+    EVP_CIPHER_CTX_cleanup(ctx);
+
+    outSize += finalEncryptedLength;
+
+    if (outSize != plainTextBlockSize)
+    {
+        result.ExtendedResultCode = ADUC_ERC_INCORRECT_BLOCK_SIZE;
+        goto done;
+    }
+
+    result.ResultCode = ADUC_GeneralResult_Success;
+done:
+
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        free(encryptedBuffer);
+        encryptedBuffer = NULL;
+        outSize = 0;
+    }
+
+    *destBuff = encryptedBuffer;
+
+    return result;
+}
+
+/**
+ * @brief Helper function that loads the @p buffToFill with the offset bytes
+ *
+ * @param buffToFill the buffer to allocated with the bytes of offset
+ * @param blockSize the blockSize for the block who's offset is being loading into the buffer
+ * @param offset the offset to be loaded into @p buffToFill
+ */
+void LoadIvBufferWithOffsetHelper(unsigned char* buffToFill, size_t blockSize, unsigned long long offset)
+{
+    unsigned long long mask = 0xFF;
+    for (int i = blockSize - 1; i >= 0; --i)
+    {
+        buffToFill[i] = mask & offset;
+        mask <<= 8; // shift left 8 bytes to select the next value
```

Comment:
dont you want a shift right?!?

Reply:
The array is [i0,i1,i2, ... , i15] where the index is each of the values.

We start the mask at the LSB at index i15 and then work our way back down to the lowest index. It would the opposite if we started out at the largest value. That's why we shift left, multiplying the mask